Remove explicit -D_POSIX_C_SOURCE global compiler options when building with picolibc

keith-packard opened the pull request:
Picolibc now has explicit Zephyr support so we no longer need _POSIX_C_SOURCE defined to get the complete Zephyr C Library API

keith-packard committed:
libc/picolibc: Remove the global definition of _POSIX_C_SOURCE

This was necessary to get Picolibc to expose the whole Zephyr C library
API, but current versions of the SDK use a version of Picolibc with
built-in Zephyr support.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
west: Update picolibc module to not define _POSIX_C_SOURCE

Make the module definitions match the toolchain definitions.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
Ideally, the one for the toolchain would depend on the version of picolibc provided, but I have no idea how to make that work.

keith-packard committed:
tests/strerror: strerror_r is not part of Zephyr C API, it's POSIX

Make sure every C library exposes strerror_r by always defining
_POSIX_C_SOURCE.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
posix: Use _POSIX_C_SOURCE=200809L instead of 200809

This constant is supposed to be defined as a long instead of an int,
presumably to support systems where int isn't large enough.

Signed-off-by: Keith Packard <keithp@keithp.com>

sylvioalves commented on drivers/wifi/esp32/src/esp_wifi_drv.c:
```diff
@@ -6,7 +6,7 @@
 #define DT_DRV_COMPAT espressif_esp32_wifi
 
-#define _POSIX_C_SOURCE 200809
+#define _POSIX_C_SOURCE 200809L
```
This can be removed as there is no non-standard posix call in Espressif's sources.

keith-packard replied:
Thanks! I'll mix that in once I've got things building.

keith-packard replied:
Looks like that worked!

keith-packard committed:
drivers/wifi/esp32: Remove _POSIX_C_SOURCE define as unneeded

This driver doesn't use any APIs outside of the Zephyr C library list, so
it doesn't need this _POSIX_C_SOURCE define.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/newlib: Add -D_POSIX_C_SOURCE=200809L to zephyr compile definitions

Newlib doesn't have Zephyr support, so the only way to get it to expose the
full Zephyr C library API is to have it expose all of the POSIX API.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests/c_lib: Remove define of _POSIX_C_SOURCE for newlib

This test does not use any APIs beyond those provided in the Zephyr
standard set, so it should not define _POSIX_C_SOURCE.

Signed-off-by: Keith Packard <keithp@keithp.com>

aescolar commented on scripts/native_simulator/Makefile:
```diff
@@ -60,7 +60,7 @@ NSI_OPT?=-O0
 NSI_WARNINGS?=-Wall -Wpedantic
 #  Preprocessor flags
-NSI_CPPFLAGS?=-D_POSIX_C_SOURCE=200809 -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED
+NSI_CPPFLAGS?=-D_POSIX_C_SOURCE=200809L -D_XOPEN_SOURCE=600 -D_XOPEN_SOURCE_EXTENDED
```
Just a sidenote: This one is not mingled with Zephyr's embedded definitions, so you should not need to change it. (it does not matter much if you do though)

keith-packard replied:
This fix aligns the code with the POSIX standard where the constants are expected to be of type long, not int. And this allows us to avoid `#undef` for code which explicitly needs POSIX functionality as the values will match.

keith-packard commented:
The compliance check failure is from the `strerror_r` test -- `strerror_r` is not part of the Zephyr C library interface, and so the test must define `_POSIX_C_SOURCE=200809L` to gain access to that symbol. I could remove this test, or we could add the function to the Zephyr API. Anyone have a preference?

aescolar commented on lib/libc/newlib/CMakeLists.txt:
```diff
@@ -73,6 +73,9 @@ if(CMAKE_C_COMPILER_ID STREQUAL "GNU")
 zephyr_link_libraries(c)
 
+# newlib needs this to expose the full Zephyr C Library API
+zephyr_compile_definitions(_POSIX_C_SOURCE=200809L)
```
I know the aim of this PR is not fixing newlib, but at least here https://github.com/zephyrproject-rtos/zephyr/pull/67052 it was not needed
Though I needed to add it in a few tests like for ex. in https://github.com/zephyrproject-rtos/zephyr/pull/67052/files#diff-afac009aa0a8f669dcbc7adb2e273455a0ba5dd4ac0fa03471801840bb987ccd
All places where I needed to add it seemed to appropriately require the macros due to the functions used.

keith-packard replied:
I believe this follows the intent of the Zephyr C library documentation -- Zephyr code should not require additional settings to gain access to any supported function, including those beyond the C standard library. For newlib, that requires exposing the entire POSIX API, at least until someone fixes newlib.

cfriedt replied:
No objection here. Hopefully we can come up with a way to do this on an as-needed basis, or at least make it specific to the platform or configured by the application.

cfriedt replied:
Darn - I guess this line might still be an issue because it changes cflags globally for the entire build, which would violate rules A.4 and likely A.5 of the [Coding Guidelines](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-4-c-standard-library-usage-restrictions-in-zephyr-kernel).

aescolar commented on tests/lib/c_lib/common/src/main.c:
```diff
@@ -15,10 +15,6 @@
  * it guarantee that ALL functionality provided is working correctly.
  */
 
-#ifdef CONFIG_NEWLIB_LIBC
```
Are you sure it is a good idea to remove this?
We will also need to set it for the host libC:
https://github.com/zephyrproject-rtos/zephyr/pull/67052/files#diff-245b0fca9de29e79ca821147f72ef70ce13d8a855bfa80e9c53b304697090f30

keith-packard replied:
Yes, the current Zephyr C library standard includes all of the functions used in this file, so there should be no additional defines required. If the C library in use does not expose the full Zephyr C library API, then that's a bug in the library integration which should be solved there and not in the test.

cfriedt replied:
Another kind of degree of freedom we wanted though was to allow sources to `#define _POSIX_C_SOURCE <ver>` at the top of the file.

It should be reasonable to leave in if that's preferred, but it might be better to do something like this

```cpp
#if !defined(_POSIX_C_SOURCE) || _POSIX_C_SOURCE < [NEEDED_VER]
#undef _POSIX_C_SOURCE
#define _POSIX_C_SOURCE [NEEDED_VER]
#endif
```

aescolar commented on tests/lib/c_lib/strerror/src/main.c:
```diff
@@ -4,9 +4,7 @@
-#ifdef CONFIG_NEWLIB_LIBC
-#define _POSIX_C_SOURCE 200809
-#endif
+#define _POSIX_C_SOURCE 200809L
```
sidenote: A benefit of #undef'ing  is that we don't need to worry about L or no L :) , so you won't get the conformance warning.

keith-packard replied:
I like leaving this alone -- it catches mistaken values used elsewhere in the system, which seems appropriate for a test.

aescolar replied:
For the c library test that is indeed fair :)

cfriedt replied:
Similarly, if needed, it could be changed to something like this
```cpp
#if !defined(_POSIX_C_SOURCE) || _POSIX_C_SOURCE < [NEEDED_VER]
#undef _POSIX_C_SOURCE
#define _POSIX_C_SOURCE [NEEDED_VER]
#endif
```

aescolar commented:
> The compliance check failure is from the `strerror_r` test -- `strerror_r` is not part of the Zephyr C library interface, and so the test must define `_POSIX_C_SOURCE=200809L` to gain access to that symbol. I could remove this test, or we could add the function to the Zephyr API. Anyone have a preference?

I guess adding it to the Zephyr API is a bit against the point (of these rules https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-5-c-standard-library-usage-restrictions-in-zephyr-codebase )
I would not have a strong objection to removing that part of the test.
Though I guess another option would be to ignore the conformance check warning? (this PR is not adding the function use, it is just exposing it)

aescolar reviewed (commented):
Some thoughts

keith-packard commented:
> I guess adding it to the Zephyr API is a bit against the point (of these rules https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#rule-a-5-c-standard-library-usage-restrictions-in-zephyr-codebase ) I would not have a strong objection to removing that part of the test. Though I guess another option would be to ignore the conformance check warning? (this PR is not adding the function use, it is just exposing it)

In fact, the PR is just adjusting the value and when it is used to conform with POSIX requirements. I'd be happy to just leave it as-is and override the conformance complaint.

keith-packard commented:
@aescolar -- looks like we might have another policy disagreement to resolve. For functions included in the Zephyr C library API, but which are not part of the C standard library API, do you believe code should need to add `#define _POSIX_C_SOURCE 200809L` before using them? I think that would be surprising to users, but perhaps you think that's correct?

aescolar commented:
> @aescolar -- looks like we might have another policy disagreement to resolve. For functions included in the Zephyr C library API, but which are not part of the C standard library API, do you believe code should need to add `#define _POSIX_C_SOURCE 200809L` before using them? I think that would be surprising to users, but perhaps you think that's correct?

(After a bit of offline discussion with @keith-packard the conclusion):
Maybe this topic could be brought to the TSC for input or others like @stephanosio could chip in for extra perspective.
My personal thinking would be that code in Zephyr should expect to need to set the same macros as in other systems when using C library extension functions. And that having that requirement would relax the requirement on the C libraries (be it newlib, or glibc, or..) or their integration to have special treatment for Zephyr.

Note that I'm not blocking this PR based on this. My input in this regard was at the comment/discussion level.

keith-packard commented:
I've done a local `twister -b` run here and found no additional problems caused by this PR.

cfriedt commented:
> The compliance check failure is from the `strerror_r` test -- `strerror_r` is not part of the Zephyr C library interface, and so the test must define `_POSIX_C_SOURCE=200809L` to gain access to that symbol. I could remove this test, or we could add the function to the Zephyr API. Anyone have a preference?

I have a PR to move it in #67001 and I have some similar PRs for POSIX Extensions to the C Standard (CX).

In this PR, it might make sense to use `#define _POSIX_C_SOURCE xxxxxL` in the test .c file (or affected areas) since they are expected to be included in `_ZEPHYR_SOURCES` IIRC.

Probably would make sense to use 
```cpp
#if !defined(_POSIX_C_SOURCE) || _POSIX_C_SOURCE < 200112L
#undef _POSIX_C_SOURCE
#define _POSIX_C_SOURCE 200112L
#endif
```

cfriedt reviewed (approved):
`200809L` works to fix the issue filed. I'm ok with the other changes too. Specifics can be sorted out later when we have some consensus maybe at an arch meeting?

cfriedt commented on arch/posix/CMakeLists.txt:
```diff
@@ -133,7 +133,7 @@ if (CONFIG_GPROF)
 endif()
 
-zephyr_compile_definitions(_POSIX_C_SOURCE=200809 _XOPEN_SOURCE=600 _XOPEN_SOURCE_EXTENDED)
+zephyr_compile_definitions(_POSIX_C_SOURCE=200809L _XOPEN_SOURCE=600 _XOPEN_SOURCE_EXTENDED)
```
This is still unfortunately a problem - even though the correct value of `_POSIX_C_SOURCE` is defined, it still sets the cflag globally, which can interfere with local cflags.

aescolar replied:
There is another PR fixing this. It does not need to be done in one go.

cfriedt commented on lib/libc/newlib/CMakeLists.txt:
```diff
@@ -73,6 +73,9 @@ if(CMAKE_C_COMPILER_ID STREQUAL "GNU")
 zephyr_link_libraries(c)
 
+# newlib needs this to expose the full Zephyr C Library API
+zephyr_compile_definitions(_POSIX_C_SOURCE=200809L)
```
Is it possible to use this instead? That would change the definition to a local one rather than a global one (which would conflict in many places).
```suggestion
zephyr_library_compile_definitions(_POSIX_C_SOURCE=200809L)
```

keith-packard replied:
No, the whole point of this define is to ensure that newlib exposes the Zephyr C API as defined by Rule A.4 in the Coding Guidelines. Unless and until newlib itself is fixed to provide Zephyr compatibility, we can either

 1. Abandon Rule A.4 and assert that Zephyr only uses ISO C interfaces
 2. Define `_POSIX_C_SOURCE=200809L` globally
 3. Find some alternative mechanism to get newlib to expose the extended functions.

aescolar replied:
@keith-packard would it be an option to only expose by default the ISO C interface, but still allow those extra functions defined in Rule A.4 (thru individual waives in PRs that need to use them, where setting the POSIX_C_SOURCE macro is allowed in particular cases).

keith-packard replied:
Are you saying that we *require* applications define `_POSIX_C_SOURCE=200809L` to gain access to the extended functions? Because if it's not required, then we're adding implicit dependencies on using only a C library with Zephyr compatibility (picolibc or the minimal C library).

That would certainly be easier for C library developers, but it seems like it will require careful auditing of changes in any file with such a definition to ensure that it doesn't accidentally start depending on additional POSIX functionality.

aescolar replied:
> Are you saying that we require ..

Yes

keith-packard replied:
It seems like that would also require allowing core Zephyr files to contain such definitions; have tooling that prevents valid code from being merged will slow down development. Because once the definition is present in a source file, no future changes to that file will see the same review, hence allowing non-approved functions to leak in. Alternatively, more sophisticated tooling that detected the use of POSIX functions beyond the approved list.

aescolar replied:
You are right that in general that would be the case, and that would be very far from ideal.

Though today a quick check seems to show:
* there is nothing in the "zephyr kernel" using `strnlen()` (the only exception in A.4).
* for more peripheral code in the tree, there is a few uses of `strnlen()` and `strtok_r()` (rule A.5 exceptions).

But given that today, at least for the tests we are running, all those built with just these 
https://github.com/zephyrproject-rtos/zephyr/pull/67040/commits/123bdfb976689e7ea73aa7917d08dd9e1b3880bb
I wonder if it would be best to, by now, postpone this particular point (I would very much like to have @stephanosio on it), and not do this particular line change (or this other commit) in this PR; The rest of the PR seems fully uncontroversial and an overall improvement.

keith-packard replied:
We can't remove instances of `#define _POSIX_C_SOURCE` in most places without somehow dealing with newlib's lack of Zephyr support. Otherwise, building code with newlib that uses `strnlen` will fail.

cfriedt reviewed (changes requested):
It would still be slightly more ideal to remove application conformance macros from global cflags.

keith-packard commented:
> It would still be slightly more ideal to remove application conformance macros from global cflags.

That requires that upstream newlib be fixed to provide Zephyr support.

keith-packard commented:
This steals @cfriedt 's idea to wrap string.h and add the non ISO-C declarations there, but is newlib specific and so doesn't risk breaking other C libaries.

aescolar commented:
Hi @keith-packard , this PR now has a very different content and target than originally. That can be quite confusing for reviewers, as the previous comments and commits are unrelated. Moreover some may have already muted this PR, and the merge criteria does not properly account for PR that morph.
The old PR content and commits and comments were relevant as they documented a discussion, and were linked from several other PRs.
May I ask you to create a new PR for the new content, push back the old commits to this branch, and close this PR so it remains as an archive of the discussion?

Setting the DNM label due to this.

cfriedt reviewed (dismissed):
This looks fine to me. Thanks @keith-packard 

Or if you prefer to make a separate PR, sure. I have no issues with the current incantation though.

Not totally necessary, but FWIR there are some CI errors right now that this will fix, so it might be a good idea to file a bug report.

CI errors are `priority: high`, so that will ensure it gets merged with the release (IIRC, we require zero high priority issues with a release).

keith-packard commented:
I've reset this PR back to the state before it became just patches for newlib compatibility. It's no longer relevant for anything but historical purposes.
